File: expansion_board/gpio/gpio.py
###############################################
#
#  file: gpio.py
#  update: 2024-08-12
#  usage: 
#      sudo python gpio.py
#
###############################################

import time
import gpiod

gpionum8 = 8
gpionum9 = 9

gpiochip6 = gpiod.Chip("6", gpiod.Chip.OPEN_BY_NUMBER)

# output set
gpio6_8 = gpiochip6.get_line(gpionum8)
gpio6_8.request(consumer="gpio", type=gpiod.LINE_REQ_DIR_OUT, default_vals=[0])

# input set
gpio6_9 = gpiochip6.get_line(gpionum9)
gpio6_9.request(consumer="gpio", type=gpiod.LINE_REQ_DIR_IN)

def main():
    try:
        while True:
            
            gpio6_8.set_value(1)
            print(gpio6_9.get_value())

            time.sleep(1)

            gpio6_8.set_value(0)
            print(gpio6_9.get_value())

            time.sleep(1)

    except Exception as e:
        
        print("exit...：", e)
    
    finally:
        
        gpio6_8.release()
        gpio6_9.release()
        
if __name__ == "__main__":  
    main()

File: expansion_board/gpio/gpio.c
/*
*
*   file: gpio.c
*   update: 2024-08-13
*   usage: 
*       sudo gcc -o gpio gpio.c -lgpiod
*       sudo ./gpio
*
*/

#include <gpiod.h>
#include <stdio.h>
#include <unistd.h>
#include <stdlib.h>
#include <signal.h> 

#define GPIOCHIP_6      "/dev/gpiochip6"

#define GPIONUM_8       (8)
#define GPIONUM_9       (9)

struct gpiod_chip *gpiochip6;        
struct gpiod_line *gpioline8;          
struct gpiod_line *gpioline9;  

static void sigint_handler(int sig_num) 
{    
    gpiod_line_release(gpioline8);
    gpiod_line_release(gpioline9);

    gpiod_chip_close(gpiochip6);

    exit(0);  
}

int main(int argc, char **argv)
{
    int ret;

    /* register exit signal ( Ctrl + c ) */
    signal(SIGINT, sigint_handler);

    /* get gpio controller */
    gpiochip6 = gpiod_chip_open(GPIOCHIP_6);  
    if(gpiochip6 == NULL)
    {
        printf("gpiod_chip_open error\n");
        return -1;
    }

    /* get gpio line */
    gpioline8 = gpiod_chip_get_line(gpiochip6, GPIONUM_8);
    if(gpioline8 == NULL)
    {
        printf("gpiod_chip_get_line error : 0\n");
        return -1;
    }

    gpioline9 = gpiod_chip_get_line(gpiochip6, GPIONUM_9);
    if(gpioline9 == NULL)
    {
        printf("gpiod_chip_get_line error : 0\n");
        return -1;
    }

    /* 设置gpio6_8为输出模式，初始电平为高 */
    ret = gpiod_line_request_output(gpioline8, "gpioline8", 1);   
    if(ret < 0)
    {
        printf("gpiod_line_request_output error : gpioline8\n");
        return -1;
    }

    /* 设置gpio6_9为输入模式*/
    ret = gpiod_line_request_input(gpioline9, "gpioline9");   
    if(ret < 0)
    {
        printf("gpiod_line_request_input error : gpioline9\n");
        return -1;
    }

    while(1)
    {   
        int bit;

        gpiod_line_set_value(gpioline8, 1);
        bit = gpiod_line_get_value(gpioline9);
        printf("%d\n", bit);

        sleep(1);

        gpiod_line_set_value(gpioline8, 0);
        bit = gpiod_line_get_value(gpioline9);
        printf("%d\n", bit);

        sleep(1);
    }

    return 0;
}